pkg/core/interop/contract/call.go

```diff
@@ -122,8 +122,44 @@ func callExFromNative(ic *interop.Context, caller util.Uint160, cs *state.Contra
 		initOff = md.Offset
 	}
 	ic.Invocations[cs.Hash]++
-	ic.VM.LoadNEFMethod(&cs.NEF, caller, cs.Hash, ic.VM.Context().GetCallFlags()&f,
-		hasReturn, methodOff, initOff, cb)
+	f = ic.VM.Context().GetCallFlags() & f
+
+	needWrap := f&(callflag.All^callflag.ReadOnly) != 0 // If the method is safe, then it's read-only and doesn't perform storage changes or emit notifications.
+	if !needWrap && ic.VM.Context().HasTryBlock() {     // If the method is not wrapped into try-catch block, then changes should be discarded anyway if exception occurs.
+		needWrap = true
+	}
+	if needWrap && ic.DAO != nil {
+		ic.DAO = ic.DAO.GetPrivate()
+	}
+	onUnload := func(commit bool, notificationsCount int) error {
+		if needWrap { // In case of CALL, CALLA, CALLL we don't need to commit/discard changes, unwrap DAO and change notificationsCount.
+			if !commit {
+				have := len(ic.Notifications)
+				if have < notificationsCount {
+					// Program error, should never happen.
+					panic(fmt.Errorf("inconsistent notifications count: should remove %d, have %d", notificationsCount, len(ic.Notifications)))
+				}
+				ic.Notifications = ic.Notifications[:have-notificationsCount]
+			}
+			if ic.DAO == nil {
+				goto callback
+			}
+			if commit {
+				_, err := ic.DAO.Persist()
+				if err != nil {
+					return fmt.Errorf("failed to persist changes %w", err)
+				}
+			}
+			ic.DAO = ic.DAO.GetUnwrapped()
```

Comment:
And you don't need `GetUnwrapped()`, you have the old DAO at the time when you create a wrapper, so just save it in the context and reuse here.

Reply:
But how about nested calls? Interop context has single state shared between all of them. Should we create a list of DAO inside the interop context?

Reply:
Finally, I got what should be done here, fixed!